src/workflow/task.rs

```diff
@@ -37,83 +45,46 @@ impl From<Closure> for Task {
-            },
+            resources: Resources::default(),
             secret: None,
         }
     }
 }
 
-impl Into<Ipld> for Task {
-    fn into(self) -> Ipld {
-        let secret_flag = match self.secret {
-            None => Ipld::Null,
-            Some(b) => Ipld::Bool(b),
-        };
-
+impl From<Task> for Ipld {
+    fn from(task: Task) -> Ipld {
         Ipld::Map(BTreeMap::from([
+            (WITH_KEY.into(), Ipld::String(task.closure.resource.into())),
+            (DO_KEY.into(), task.closure.action.into()),
+            (INPUTS_KEY.into(), task.closure.inputs.into()),
+            (RESOURCES_KEY.into(), task.resources.into()),
             (
-                "with".to_string(),
-                Ipld::String(self.closure.resource.into()),
+                SECRETS_KEY.into(),
+                task.secret.map(Ipld::Bool).unwrap_or(Ipld::Null),
             ),
-            ("do".to_string(), self.closure.action.into()),
-            ("inputs".to_string(), self.closure.inputs.into()),
-            ("resources".to_string(), self.resources.into()),
-            ("secret".to_string(), secret_flag),
         ]))
     }
 }
 
 impl TryFrom<Ipld> for Task {
-    type Error = ();
+    type Error = anyhow::Error;
 
     fn try_from(ipld: Ipld) -> Result<Self, Self::Error> {
-        match ipld {
-            Ipld::Map(ref assoc) => {
-                let res = match assoc.get("resources") {
-                    Some(v) => v.clone(),
-                    _ => Ipld::Map(BTreeMap::new()),
-                };
+        let map = ipld_serde::from_ipld::<BTreeMap<String, Ipld>>(ipld.clone())?;
 
-                let fuel: Option<u32> =
-                    res.get("fuel")
-                        .map_err(|_| ())
-                        .and_then(|ipld_fuel| match ipld_fuel {
-                            Ipld::Integer(int) => Ok(u32::try_from(*int).ok()),
-                            _ => Err(()),
-                        })?;
-
-                let time: Option<u32> =
-                    res.get("time")
-                        .map_err(|_| ())
-                        .and_then(|ipld_fuel| match ipld_fuel {
-                            Ipld::Integer(int) => Ok(u32::try_from(*int).ok()),
-                            _ => Err(()),
-                        })?;
-
-                let metadata: Ipld = match assoc.get("meta") {
-                    Some(ipld) => ipld.clone(),
-                    None => Ipld::Null,
-                };
-
-                // Is it secret? Is it safe?!
-                let secret: Option<bool> =
-                    assoc.get("secret").ok_or(()).and_then(|ipld| match ipld {
-                        Ipld::Bool(b) => Ok(Some(*b)),
-                        Ipld::Null => Ok(None),
-                        _ => Err(()),
-                    })?;
+        let resources = if let Some(ipld) = map.get(RESOURCES_KEY) {
+            Resources::try_from(ipld)?
+        } else {
+            Resources::default()
+        };
```

Comment:
`let resources = if let Some(ipld) =`

Woah, that's gnarly... but idiomatic?

Could it be rephrased as something like...

```rust
let resources = map
    .get(RESOURCES_KEY)
    .map(Resources::try_from)
    .unwrap_or_else(Resources::default);
```

...or is that not commonly done?

Reply:
@expede yep, this is better (had this in the about to push)

Reply:
even cleaner:

```rust
let resources = map
            .get(RESOURCES_KEY)
            .map_or_else(|| Ok(Resources::default()), Resources::try_from)?;
```